cli/exec.go

```diff
@@ -0,0 +1,192 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package cli
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+	"time"
+
+	"github.com/Azure/acr-builder/builder"
+	"github.com/Azure/acr-builder/cmder"
+	"github.com/Azure/acr-builder/graph"
+	"github.com/Azure/acr-builder/templating"
+	"github.com/Azure/acr-builder/volume"
+	"github.com/google/uuid"
+	"github.com/spf13/cobra"
+)
+
+const execLongDesc = `
+This command can be used to execute a pipeline.
+`
+
+type execCmd struct {
+	out       io.Writer
+	dryRun    bool
+	stepsFile string
+
+	registry         string
+	registryUserName string
+	registryPassword string
+
+	// Build-time parameters for rendering
+	templatePath     string
+	templateValues   []string
+	buildID          string
+	buildCommit      string
+	buildTag         string
+	buildRepository  string
+	buildBranch      string
+	buildTriggeredBy string
+}
+
+func newExecCmd(out io.Writer) *cobra.Command {
+	e := &execCmd{
+		out: out,
+	}
+
+	cmd := &cobra.Command{
+		Use:   "exec",
+		Short: "Execute a pipeline",
+		Long:  execLongDesc,
+		RunE:  e.run,
+	}
+
+	f := cmd.Flags()
+
+	f.StringVar(&e.stepsFile, "steps", "", "the steps file to use when building")
+
+	f.StringVarP(&e.registry, "registry", "r", "", "the name of the registry")
+	f.StringVarP(&e.registryUserName, "username", "u", "", "the username to use when logging into the registry")
+	f.StringVarP(&e.registryPassword, "password", "p", "", "the password to use when logging into the registry")
+
+	// TODO: better names and shorthand
+	// Rendering parameters
+	f.StringVar(&e.buildID, "id", "", "the build ID")
+	f.StringVarP(&e.buildCommit, "commit", "c", "", "the commit SHA")
+	f.StringVarP(&e.buildTag, "tag", "t", "", "the build tag")
+	f.StringVar(&e.buildRepository, "repository", "", "the build repository")
+	f.StringVarP(&e.buildBranch, "branch", "b", "", "the build branch")
+	f.StringVar(&e.buildTriggeredBy, "triggered-by", "", "what the build was triggered by")
+	f.StringVar(&e.templatePath, "template-path", "", "the path to the job to render")
+	f.StringArrayVar(&e.templateValues, "set", []string{}, "set values on the command line (use `--set` multiple times or use commas: key1=val1,key2=val2)")
+	f.BoolVar(&e.dryRun, "dryrun", false, "evaluates the pipeline but doesn't execute it")
```

Comment:
>dryrun [](start = 23, length = 6)

dry-run, can build also have a todo item to support dry-run?

Reply:
added support for this